content/courses/advanced-ada/chapters/statements.rst

```diff
@@ -3,6 +3,133 @@ Statements
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Statement_Identifier is
+       pragma Warnings (Off, "is not referenced");
+    begin
+       <<Show_Hello>> Put_Line ("Hello World!");
+       <<Show_Test>>  Put_Line ("This is a test.");
+
+       <<Show_Separator>>
+       <<Show_Block_Separator>>
+       Put_Line ("====================");
+    end Show_Statement_Identifier;
+
+Here, we're labeling each statement. For example, we use the :ada:`Show_Hello`
+label to identify the :ada:`Put_Line ("Hello World!");` statement. Note that we
+can use multiple labels a single statement. In this code example, we use the
+:ada:`Show_Separator` and :ada:`Show_Block_Separator` labels for the same
+statement.
+
+We can use labels with compound statements as well. For example, we can label
+a :ada:`for` loop:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Statements.Statement_Identifier_2
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Statement_Identifier is
+       pragma Warnings (Off, "is not referenced");
+
+       Arr   : constant array (1 .. 5) of Integer := (1, 4, 6, 42, 49);
+       Found : Boolean := False;
+    begin
+       <<Find_42>> for E of Arr loop
+          if E = 42 then
+             Found := True;
+             exit;
+          end if;
+       end loop;
+
+       Put_Line ("Found: " & Found'Image);
+    end Show_Statement_Identifier;
+
+.. admonition:: Relevant topics
+
+    In addition to labels, loops and block statements allow us to use a
+    statement identifier. In simple terms, instead of writing
+    :ada:`<<Some_Label>>`, we write :ada:`Some_Label :`.
+
+    We could rewrite the previous code example using a loop statement
+    identifier:
+
+    .. code:: ada run_button project=Courses.Advanced_Ada.Statements.Loop_Statement_Identifier
+
+        with Ada.Text_IO; use Ada.Text_IO;
+
+        procedure Show_Statement_Identifier is
+           Arr   : constant array (1 .. 5) of Integer := (1, 4, 6, 42, 49);
+           Found : Boolean := False;
+        begin
+           Find_42 : for E of Arr loop
+              if E = 42 then
+                 Found := True;
+                 exit Find_42;
+              end if;
+           end loop Find_42;
+
+           Put_Line ("Found: " & Found'Image);
+        end Show_Statement_Identifier;
+
+    Later in this chapter, we discuss this topic in more details.
+
+Although the usage of :ada:`goto` statements should be avoided as much as
+possible, it's worth mentioning that we can use a :ada:`goto label;` statement
+to *jump* to a specific label. For example:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Statements.Statement_Identifier_3
+
+    procedure Show_Cleanup is
+       Some_Error : Boolean;
+    begin
+       Some_Error := False;
+
+       if Some_Error then
+          goto Cleanup;
+       end if;
+
+       <<Cleanup>> null;
+    end Show_Cleanup;
+
+Here, we transfer the control to the *cleanup* statement as soon as an error is
```

Comment:
And here I'd suggest a description of using a goto statement in a loop to implement a continue statement, which can simplify the number of boolean flags required for loop termination in some cases.

Reply:
Usually, I hardly need `goto` statements, so I'm not sure I've ever encountered this use case where a `goto` would be preferable in a loop. Could you please show me what code you have in mind?

Reply:
I've only used it once, in an executive:   <<spin>> goto spin;

They are rife throughout the gnat compiler code. The simplest I could find is this one:

   procedure Qualify_All_Entity_Names is
      E   : Entity_Id;
      Ent : Entity_Id;
      Nod : Node_Id;

   begin
      for J in Name_Qualify_Units.First .. Name_Qualify_Units.Last loop
         Nod := Name_Qualify_Units.Table (J);

         --  When a scoping construct is ignored Ghost, it is rewritten as
         --  a null statement. Skip such constructs as they no longer carry
         --  names.

         if Nkind (Nod) = N_Null_Statement then
            goto Continue;
         end if;

         E := Defining_Entity (Nod);
         Reset_Buffers;
         Qualify_Entity_Name (E);

         --  Normally entities in the qualification list are scopes, but in the
         --  case of a library-level package renaming there is an associated
         --  variable that encodes the debugger name and that variable is
         --  entered in the list since it occurs in the Aux_Decls list of the
         --  compilation and doesn't have a normal scope.

         if Ekind (E) /= E_Variable then
            Ent := First_Entity (E);
            while Present (Ent) loop
               Reset_Buffers;
               Qualify_Entity_Name (Ent);
               Next_Entity (Ent);

               --  There are odd cases where Last_Entity (E) = E. This happens
               --  in the case of renaming of packages. This test avoids
               --  getting stuck in such cases.

               exit when Ent = E;
            end loop;
         end if;

         <<Continue>>
         null;
      end loop;
   end Qualify_All_Entity_Names;

However the better case is in while-loops, where "implementing" a continue statement via the goto allows one not need another boolean variable to control the loop iteration. 

All that said, I think refactoring to simplify the loop content is a better approach. I only mention the "continue statement" stuff in case an example is needed.

Reply:
Thanks, now I understand what you mean.

I just wrote a small subsection that explains the basic idea behind a `Continue` label.